File: Android/app/src/main/java/dev/dikka/penferry/PenPacket.kt
package dev.dikka.penferry

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress

enum class PenEvent(val id: Byte) {
    HOVER_MOVE(0x10),
    HOVER_EXIT(0x11),
    CONTACT_MOVE(0x20),
    CONTACT_DOWN(0x22),
    CONTACT_UP(0x21),
    SUPP_ACTION(0x0F),
    DUMMY(0x00);

    companion object {
        fun find(value: Byte): PenEvent? = values().find { it.id == value }
    }
}

data class PenPacket(val event: PenEvent, val arg1: Float = 0f, val arg2: Float = 0f, val arg3: Float = 0f, val buttonPressed: Boolean = false) {
    companion object {
        val size = PenPacket(PenEvent.DUMMY).encode().size
    };
    override fun toString(): String {
        return "${this.event.name}, ${this.arg1}, ${this.arg2}, ${this.arg3}, ${this.buttonPressed}"
    }
}

fun PenPacket.send(address: String, port: Int) {
    CoroutineScope(Dispatchers.IO).launch {
        withContext(Dispatchers.IO) {
            DatagramSocket().use { socket ->
                socket.broadcast = true
                val data = this@send.encode()
                socket.send(DatagramPacket(data, data.size, InetAddress.getByName(address), port))
                output + "${this@send}"
            }
        }
    }
}

fun PenPacket.encode(): ByteArray {
    return byteArrayOf(this.event.id) + arg1.toBytes() + arg2.toBytes() + arg3.toBytes() + buttonPressed.toByte()
}

fun Float.toBytes(): ByteArray {
    val buffer = ByteArray(4)
    for (i in 0..3) buffer[i] = (this.toRawBits() shr (i * 8)).toByte()
    return buffer
}

fun ByteArray.toFloat(): Float {
    return Float.fromBits(
        (this[3].toInt() shl 24) or
                (this[2].toInt() and 0xff shl 16) or
                (this[1].toInt() and 0xff shl 8) or
                (this[0].toInt() and 0xff)
    )
}

fun Boolean.toByte(): Byte = if (this) 0x01 else 0x00

File: Windows/internal/penpacket/penpacket.go
package penpacket

import (
	"encoding/binary"
	"fmt"
	"math"
)

type PenEvent byte

const (
	HOVER_MOVE  PenEvent = 0x10
	HOVER_EXIT  PenEvent = 0x11
	CONTACT_MOVE PenEvent = 0x20
	CONTACT_DOWN PenEvent = 0x22
	CONTACT_UP   PenEvent = 0x21
    SUPP_ACTION  PenEvent = 0x0F
	DUMMY        PenEvent = 0x00
)

func (pe PenEvent) String() string {
	switch pe {
	case HOVER_MOVE:
		return "HOVER_MOVE"
	case HOVER_EXIT:
		return "HOVER_EXIT"
	case CONTACT_MOVE:
		return "CONTACT_MOVE"
	case CONTACT_DOWN:
		return "CONTACT_DOWN"
	case CONTACT_UP:
		return "CONTACT_UP"
    case SUPP_ACTION:
        return "SUPP_ACTION"
	case DUMMY:
		return "DUMMY"
	default:
		return "UNKNOWN_EVENT"
	}
}

type PenPacket struct {
	Event         PenEvent
	Arg1, Arg2, Arg3 float32
	ButtonPressed bool
}

func (pp *PenPacket) String() string {
	switch pp.Event {
	case HOVER_MOVE:
		return fmt.Sprintf("%s{x: %v, y: %v, %v}", pp.Event, pp.Arg1, pp.Arg2, pp.ButtonPressed)
    case CONTACT_MOVE:
		return fmt.Sprintf("%s{x: %v, y: %v, %v, %v}", pp.Event, pp.Arg1, pp.Arg2, pp.Arg3, pp.ButtonPressed)
	case HOVER_EXIT, CONTACT_DOWN, CONTACT_UP, SUPP_ACTION, DUMMY:
		return fmt.Sprintf("%s{}", pp.Event)
	default:
		return "Unknown Event"
	}
}

func Decode(data []byte) (*PenPacket, error) {
	if len(data) != 14 {
		return nil, fmt.Errorf("Invalid data length. Expected 14, got %v", len(data))
	}

	pp := &PenPacket{
		Event:         PenEvent(data[0]),
		Arg1:          bytesToFloat32(data[1:5]),
		Arg2:          bytesToFloat32(data[5:9]),
		Arg3:          bytesToFloat32(data[9:13]),
		ButtonPressed: data[13] == 0x01,
	}
	
	return pp, nil
}

func bytesToFloat32(b []byte) float32 {
	bits := binary.LittleEndian.Uint32(b)
	float := math.Float32frombits(bits)
	return float
}
